06x-new-kafka-roller.md

```diff
@@ -0,0 +1,350 @@
+ | State            | Description      | Possible transitions |
+ | :--------------- | :--------------- | :----------- |
+ | UNKNOWN               | The initial state when creating `Context` for a node. We expect to transition from this state fairly quickly after creating the context for nodes.  | `NOT_RUNNING` `NOT_READY` `RECOVERING` `SERVING` |
+ | NOT_RUNNING           | Node is not running (Kafka process is not running). This is determined via Kubernetes API, more details for it below. | `NOT_READY`, `NOT_RUNNING`. | `RESTARTED` `SERVING` | 
+ | NOT_READY             | Node is running but not ready to serve requests which is determined by Kubernetes readiness probe (broker state < 2 OR == 127 OR controller is not listening on port). | `RESTARTED` `SERVING` |
+ | RESTARTED             | After successful `kubectl delete pod`. | `NOT_RUNNING` `NOT_READY` `RECOVERING` `SERVING` |
+ | RECONFIGURED          | After successful Kafka node config update via Admin client. | `NOT_RUNNING` `NOT_READY` `RESTARTED` `SERVING` |
+ | RECOVERING            | Node has started but is in log recovery (broker state == 2). | `SERVING` |
+ | SERVING               | Node is in running state and ready to serve requests which is determined by Kubernetes readiness probe (broker state >= 3 AND != 127 OR controller is listening on port). | `RESTARTED` `RECONFIGURED` `LEADING_ALL_PREFERRED` |
+ | LEADING_ALL_PREFERRED | Node is in running state and leading all preferred replicas. | None
+
+The definitions of broker states can be found via the following link: [Broker States](https://github.com/apache/kafka/blob/3.7/metadata/src/main/java/org/apache/kafka/metadata/BrokerState.java).
+
+The definitions of the possible restart reasons can be found via the following link: [Restart Reasons](https://github.com/strimzi/strimzi-kafka-operator/blob/0.40.0/cluster-operator/src/main/java/io/strimzi/operator/cluster/model/RestartReason.java)
+
+#### NOT_RUNNING state
+
+If one of the following is true, then node's state is `NOT_RUNNING:
+- no pod exists for this node
+- unable to get the `Pod Status` for the pod 
+- the pod has `Pending` status with `Unschedulable` reason
+- the pod has container status `ContainerStateWaiting` with `CrashLoopBackOff` or `ImagePullBackOff` reason
+If none of the above is true but the node is not ready, then its state would be `NOT_READY`.
+
+#### Flow diagram describing the overall flow of the states
+![The new roller flow](./images/06x-new-roller-flow.png)
+
+
+### Configurability
+The following can be the configuration options for the new KafkaRoller:
+
+| Configuration          | Default value | Exposed to user | Description                                                                                                                                                                                                                                                   |
```

Comment:
general question on the "Exposed to user" column. What's the rational behind deciding to not expose some of them to the user? Could all of them helping if users are dealing with different environments, different latency, network conditions, where changing all of them could help?

Reply:
We could if we think that's useful. We followed what the current roller does, it doesn't expose how many attempts are allowed for each pod (maxAttempts is set to 10 by the reconciler).